### DISCUSSION TITLE
Generate multiple urls

### DISCUSSION THREAD

**Discussion by boxadesign** [GoogleChrome/lighthouse-ci#611] [Q&A]
Hi I want to run lighthouse ci on all pages in a static site generator. The below doesn't work. I need to wait for all of the page paths to be returned as an array (`urls`) and then passed in as the value to `url` in the lighthouserc config. When I run `lhci autorun` from the command line it's not picking up the `ci: {}` config.

```
const getPages = require('./lighthouse');

module.exports = (async () => {
  const urls = await getPages();
  const obj = {
    ci: {
      collect: {
        url: urls,
        numberOfRuns: 1,
      },
      preset: 'lighthouse:no-pwa',
      assertions: {
        'categories:accessibility': ['warn', { minScore: 0.9 }],
      },
      upload: {
        target: 'temporary-public-storage',
      },
    },
  };
  return obj;
})().catch(err => {
  console.error(err);
});
```

**Comment by patrickhulce** [comment]
Thanks for filing @boxadesign!

The `.lighthouserc.js` config file cannot return a promise; it *must* return the configuration object. I'd suggest doing whatever async work you'd like *before* you run `lhci autorun` and store the results in a file or script of some sort. This is actually what LHCI does to run on itself.

https://github.com/GoogleChrome/lighthouse-ci/blob/ab9f5c3f6c3704f67cf94ef749d9c28f1652f9a0/scripts/ci-dogfood.sh#L16-L18

**Comment by boxadesign** [comment]
Thanks alot. I've now implemented this and it works great.  Pasted below to see if there are any improvements you can suggest.

I'm running lighthouse tests on all components in a large design system. Is there a way to run these in parallel with multiple chrome instances?

```
#!/bin/bash
node ./lighthouse/lighthouse-get-pages.js
npx http-server -p 9000 ./build &
npm install -g @lhci/cli@0.7.x
lhci healthcheck --fatal

for url in $(jq '.urls[]' ./lighthouse/urls.json); do
    lhci collect "--url=$url" --config=./lighthouse/lighthouserc.js || exit 1
done
EXIT_CODE=$?

kill $!
exit $EXIT_CODE
```


Thanks.

**Reply by patrickhulce** [reply]
Is that working for you? I wouldn't expect any URL but the last one to be saved without the `--additive` flag.

> Is there a way to run these in parallel with multiple chrome instances?

Not if you care about the performance category. If you don't then sure, spin up multiple `lhci collect` jobs at once!

**Reply by boxadesign** [reply]
Hi thanks and you're right I did need the `--additive` flag,

Here is what I have got to...

```
#!/bin/bash
node ./lighthouse/lighthouse-get-pages.js
npx http-server -p 9000 ./build &
npm install -g @lhci/cli@0.7.x
lhci healthcheck --fatal
for url in $(jq '.urls[]' ./lighthouse/urls.json); do
    lhci collect "--url=$url" --additive --config=./lighthouse/lighthouserc.js
done
lhci assert --config=./lighthouse/lighthouserc.js
lhci upload --config=./lighthouse/lighthouserc.js

EXIT_CODE=$?

kill $!
exit $EXIT_CODE
```

Could you offer any advice on how I would spin up multiple `lhci collect` jobs at once? 

Here is a segment of my `lighthouse-ci.yml`...

```
      - name: Run Lighthouse CI
        run: ./lighthouse/lighthouse.sh
        env:
          LHCI_GITHUB_APP_TOKEN: ${{ secrets.LHCI_GITHUB_APP_TOKEN }}
```

This code runs through all pages and sends a link to each report for each page to github status checks. Which I think will provide what we need.

**Reply by patrickhulce** [reply]
Heads up you probably want to using the exit code from `assert`, not `upload`.

https://unix.stackexchange.com/questions/242087/run-commands-in-parallel-and-wait-for-one-group-of-commands-to-finish-before-sta might help you on the parallelization

**Reply by boxadesign** [reply]
Thanks for the advice. This assisted me more in running in parallel. https://unix.stackexchange.com/questions/103920/parallelize-a-bash-for-loop

```
#!/bin/bash
node ./lighthouse/lighthouse-get-pages.js
npx http-server -p 9000 ./build &
npm install -g @lhci/cli@0.7.x
lhci healthcheck --fatal

for url in $(jq '.urls[]' ./lighthouse/urls.json); do sem
    lhci collect "--url=$url" --additive --config=./lighthouse/lighthouserc.js &
done
sem --wait
lhci assert --config=./lighthouse/lighthouserc.js
EXIT_CODE=$?
lhci upload --config=./lighthouse/lighthouserc.js
kill $!
exit $EXIT_CODE
```
